File: java/json-transform/src/main/java/co/nlighten/jsontransform/template/TemplateParameter.java
package co.nlighten.jsontransform.template;

import co.nlighten.jsontransform.*;
import co.nlighten.jsontransform.adapters.JsonAdapter;

public class TemplateParameter {
    private final String name;
    private String mDefault;

    String getDefault() {
        return mDefault;
    }

    void setDefault(String value) {
        mDefault = value;
    }

    TemplateParameter(String name, String defaultValue) {
        this.name = name;
        mDefault = defaultValue == null ? "" : defaultValue;
    }

    String getStringValue(ParameterResolver resolver, JsonAdapter<?,?,?> adapter) {
        if (resolver == null) {
            return null;
        }

        var val = resolver.get(name);

        if (name.startsWith("$$") && (name.equals(val) || val == null)) {
            var x = new TextTemplateJsonTransformer(adapter);
            val = x.transformString(adapter.wrap(name), resolver);
        }

        return val == null || adapter.isNull(val)
                ? mDefault
                : adapter.getAsString(val).replace("{", "\\{");
    }
}

File: javascript/json-transform/src/template/TemplateParameter.ts
import { ParameterResolver } from "../ParameterResolver";
import { getCurrentJsonTransformFunction } from "./TextTemplateJsonTransformFunction";

export default class TemplateParameter {
  private readonly name: string;
  private mDefault: string;

  getDefault() {
    return this.mDefault;
  }

  setDefault(value: string) {
    this.mDefault = value;
  }

  constructor(name: string, defaultValue: string) {
    this.name = name;
    this.mDefault = defaultValue == null ? "" : defaultValue;
  }

  async getStringValue(resolver: ParameterResolver) {
    if (resolver == null) {
      return null;
    }
    let val = resolver.get(this.name);

    if (this.name.startsWith("$$") && (this.name === val || val == null)) {
      val = await getCurrentJsonTransformFunction().call(null, this.name, resolver);
    }

    return val == null ? this.mDefault : val.replace(/\{/g, "\\{");
  }
}
